File: app/src/main/java/com/stackanswer/viewmodel/factory/DetailShowViewModelFactory.java
package com.stackanswer.viewmodel.factory;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.stackanswer.source.repository.DetailShowRepository;
import com.stackanswer.injection.Injection;
import com.stackanswer.viewmodel.DetailShowViewModel;

public class DetailShowViewModelFactory extends ViewModelProvider.NewInstanceFactory {
    private static volatile DetailShowViewModelFactory INSTANCE;

    private final DetailShowRepository mDetailShowRepository;

    private DetailShowViewModelFactory(DetailShowRepository academyRepository) {
        mDetailShowRepository = academyRepository;
    }

    public static DetailShowViewModelFactory getInstance(Context context) {
        if (INSTANCE == null) {
            synchronized (DetailShowViewModelFactory.class) {
                if (INSTANCE == null) {
                    INSTANCE = new DetailShowViewModelFactory(Injection.provideDetailShowRepository(context));
                }
            }
        }
        return INSTANCE;
    }

    @SuppressWarnings("unchecked")
    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {

        if (modelClass.isAssignableFrom(DetailShowViewModel.class)) {
            return (T) new DetailShowViewModel(mDetailShowRepository);
        }
        throw new IllegalArgumentException("Unknown ViewModel class: " + modelClass.getName());
    }
}

File: app/src/main/java/com/stackanswer/viewmodel/factory/kotlin/DetailShowViewModelFactory.kt
package com.stackanswer.viewmodel.factory.kotlin

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.stackanswer.core.domain.usecase.DetailShowUseCaseKt
import com.stackanswer.injection.kotlin.Injection
import com.stackanswer.viewmodel.DetailShowViewModelKt

class DetailShowViewModelFactory private constructor(private val tourismUseCase: DetailShowUseCaseKt) :
    ViewModelProvider.NewInstanceFactory() {

    companion object {
        @Volatile
        private var instance: DetailShowViewModelFactory? = null

        fun getInstance(context: Context): DetailShowViewModelFactory =
            instance ?: synchronized(this) {
                instance ?: DetailShowViewModelFactory(
                    Injection.provideDetailShowUseCase(context)
                )
            }
    }

    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T =
        when {
            modelClass.isAssignableFrom(DetailShowViewModelKt::class.java) -> {
                DetailShowViewModelKt(tourismUseCase) as T
            }
            else -> throw Throwable("Unknown ViewModel class: " + modelClass.name)
        }
}